From 546c29e9d1161f829074d0b11dde2eb168ca2bba Mon Sep 17 00:00:00 2001
From: Geoffrey Thomas <geofft@ldpreload.com>
Date: Tue, 27 Jul 2021 18:58:12 -0400
Subject: [PATCH 17/22] closes bpo-44751: Move crypt.h include from public
 header to _cryptmodule (GH-27394)

Automerge-Triggered-By: GH:benjaminp
(cherry picked from commit 196998e220d6ca030e5a1c8ad63fcaed8e049a98)
---
 Include/Python.h                                    | 13 -------------
 .../C API/2021-07-27-17-29-12.bpo-44751.4qmbDG.rst  |  1 +
 Modules/_cryptmodule.c                              |  3 +++
 3 files changed, 4 insertions(+), 13 deletions(-)
 create mode 100644 Misc/NEWS.d/next/C API/2021-07-27-17-29-12.bpo-44751.4qmbDG.rst

diff --git a/Include/Python.h b/Include/Python.h
index d6e5b139ac..0bc36c23d8 100644
--- a/Include/Python.h
+++ b/Include/Python.h
@@ -35,19 +35,6 @@
 #ifndef MS_WINDOWS
 #include <unistd.h>
 #endif
-#ifdef HAVE_CRYPT_H
-#if defined(HAVE_CRYPT_R) && !defined(_GNU_SOURCE)
-/* Required for glibc to expose the crypt_r() function prototype. */
-#  define _GNU_SOURCE
-#  define _Py_GNU_SOURCE_FOR_CRYPT
-#endif
-#include <crypt.h>
-#ifdef _Py_GNU_SOURCE_FOR_CRYPT
-/* Don't leak the _GNU_SOURCE define to other headers. */
-#  undef _GNU_SOURCE
-#  undef _Py_GNU_SOURCE_FOR_CRYPT
-#endif
-#endif
 
 /* For size_t? */
 #ifdef HAVE_STDDEF_H
diff --git a/Misc/NEWS.d/next/C API/2021-07-27-17-29-12.bpo-44751.4qmbDG.rst b/Misc/NEWS.d/next/C API/2021-07-27-17-29-12.bpo-44751.4qmbDG.rst
new file mode 100644
index 0000000000..d7b9f09819
--- /dev/null
+++ b/Misc/NEWS.d/next/C API/2021-07-27-17-29-12.bpo-44751.4qmbDG.rst	
@@ -0,0 +1 @@
+Remove ``crypt.h`` include from the public ``Python.h`` header.
diff --git a/Modules/_cryptmodule.c b/Modules/_cryptmodule.c
index 5d03f45f64..95e4a90f0c 100644
--- a/Modules/_cryptmodule.c
+++ b/Modules/_cryptmodule.c
@@ -4,6 +4,9 @@
 #include "Python.h"
 
 #include <sys/types.h>
+#ifdef HAVE_CRYPT_H
+#include <crypt.h>
+#endif
 
 /* Module crypt */
 
-- 
2.35.1

